cnd/src/storage/http_api.rs

```diff
@@ -0,0 +1,497 @@
+        &self,
+        swap_id: LocalSwapId,
+    ) -> anyhow::Result<
+        BobSwap<asset::Bitcoin, asset::Erc20, hbit::FinalizedAsRedeemer, herc20::Finalized>,
+    > {
+        let alpha_state = self.hbit_states.get(&swap_id).await?;
+        let beta_state = self.herc20_states.get(&swap_id).await?;
+
+        let (_, herc20, hbit, secret_hash) = self.load_tp_herc20_hbit(swap_id).await?;
+        let hbit_asset = hbit.amount.0.into();
+
+        let swap = match (alpha_state, beta_state) {
+            (Some(alpha_state), Some(beta_state)) => {
+                let beta_finalized = Herc20WithState {
+                    herc20,
+                    state: beta_state,
+                }
+                .try_into()?;
+
+                BobSwap::Finalized {
+                    alpha_finalized: hbit::FinalizedAsRedeemer {
+                        asset: hbit_asset,
+                        network: hbit.network.0.into(),
+                        transient_redeem_identity: self
+                            .seed
+                            .derive_swap_seed(swap_id)
+                            .derive_transient_redeem_identity(),
+                        transient_refund_identity: hbit
+                            .transient_identity
+                            .ok_or(db::Error::IdentityNotSet)?
+                            .0,
+                        final_redeem_identity: hbit.final_identity.0,
+                        expiry: hbit.expiry.0.into(),
+                        state: alpha_state,
+                    },
+                    beta_finalized,
+                    secret_hash: secret_hash
+                        .ok_or(db::Error::SecretHashNotSet)?
+                        .secret_hash
+                        .0,
+                }
+            }
+            _ => BobSwap::Created {
+                alpha_created: hbit_asset,
+                beta_created: herc20.into(),
+            },
+        };
+
+        Ok(swap)
+    }
+}
+
+struct Herc20WithState {
+    herc20: Herc20,
+    state: herc20::State,
+}
+
+impl TryFrom<Herc20WithState> for herc20::Finalized {
+    type Error = db::Error;
+
+    fn try_from(from: Herc20WithState) -> Result<Self, Self::Error> {
+        let herc20 = from.herc20;
+        let state = from.state;
+
+        let herc20_asset = asset::Erc20 {
+            quantity: herc20.amount.0.into(),
+            token_contract: herc20.token_contract.0.into(),
+        };
+
+        Ok(herc20::Finalized {
+            asset: herc20_asset,
+            chain_id: herc20.chain_id.0.into(),
+            refund_identity: herc20
+                .refund_identity
+                .ok_or(db::Error::IdentityNotSet)?
+                .0
+                .into(),
+            redeem_identity: herc20
+                .redeem_identity
+                .ok_or(db::Error::IdentityNotSet)?
+                .0
+                .into(),
+            expiry: herc20.expiry.0.into(),
+            state,
+        })
+    }
+}
```

Comment:
This could just be a function analogous to `build_xyz_params` then you don't need the rather awkwardly named abstraction of `Herc20WithState` :)

Reply:
> What I really wanted to do was implement this trait on a tuple - seems we can't do that in Rust ATM though.

Haha, I always try to do this :(